src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,429 @@
+        }
+
+        private void CreateContainerClient()
+        {
+            var tokenCredential = EnsureArg.IsNotNull(_partitionLockingOptions.StorageTokenCredential, nameof(_partitionLockingOptions.StorageTokenCredential));
+            _containerClient = new BlobContainerClient(_partitionLockingOptions.BlobContainerUri, tokenCredential.GetCredential());
+            _containerClient.CreateIfNotExists();
+        }
+
+        private async Task RegisterProcessor()
+        {
+            // write process id to storage
+            _logger.LogTrace($"Processor Id: {_processorId}");
+            BlobClient blobClient = _containerClient.GetBlobClient($"activepods/{_processorId}");
+
+            var activeUntil = new Dictionary<string, string>()
+            {
+                { "activeuntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+            };
+
+            BlobUploadOptions options = new BlobUploadOptions() { Metadata = activeUntil };
+            await blobClient.UploadAsync(BinaryData.FromString(string.Empty), options);
+
+            _logger.LogTrace($"Registered processor {_processorId}");
+        }
+
+        private async Task RegisterProcessorLoop()
+        {
+            while (true)
+            {
+                await RegisterProcessor();
+                await Task.Delay(15000);
+            }
+        }
+
+        private async Task<string[]> GetPartitionsIdsForEventHub()
+        {
+            await using (var consumer = new EventHubConsumerClient(
+                _eventHubClientOptions.EventHubConsumerGroup,
+                _eventHubClientOptions.EventHubNamespaceFQDN,
+                _eventHubClientOptions.EventHubName,
+                _eventHubClientOptions.EventHubTokenCredential))
+            {
+                return await consumer.GetPartitionIdsAsync();
+            }
+        }
+
+        private async Task AcquirePartitions(CancellationToken ct)
+        {
+            // get all the processors in active status
+            var activeProcessors = ListBlobs(_containerClient, "activepods");
+            var activeProcessorCount = await GetActiveProcessors(activeProcessors, _containerClient, ct);
+            _totalProcessorsRunning = activeProcessorCount;
+            _logger.LogTrace($"Active processor count: {activeProcessorCount}");
+
+            if (activeProcessorCount == 0)
+            {
+                throw new Exception("active processors is 0");
+            }
+
+            // divide the partitions among the active pods
+            // todo: handle cases where there is a remainder
+            // for example 8 partitions and 6 pods, which would require 4 pods to have 1 partition, and 2 pods to have 2 partitions
+            // for now, we will round down
+
+            _logger.LogTrace($"The Event Hub has {_eventHubPartitions.Length} partitions");
+
+            var suggestedPartitionsPerProcessor = Math.Floor((double)_eventHubPartitions.Length / activeProcessorCount);
```

Comment:
yeah I agree it would be better in scenarios where the number of partitions is not evenly divided by the number of pods. I will look into adding this today

Reply:
Still a a solution for this, if there is a remainder, the remainder will go unclaimed.